swarm/animator.py

```diff
@@ -45,3 +59,306 @@ def animate(i):
+        return cls(hist_data, num_bins, [_kwargs_hook(**kwargs)])
+
+    def init(self, ax: plt.Axes):
+        """Copied from https://matplotlib.org/3.1.1/gallery/animation/animated_histogram.html"""
+        import matplotlib.patches as patches
+        import matplotlib.path as path
+
+        self.artists = []
+        ax.clear()
+        _apply_hook(ax, self.hook)
+        init_data = self.data.flatten()
+        n, bins = np.histogram(init_data, self.num_bins)
+
+        # get the corners of the rectangles for the histogram
+        left = np.array(bins[:-1])
+        right = np.array(bins[1:])
+        self._bottom = np.zeros(len(left))
+        top = self._bottom + n / (self.num_frames * 0.9)
+        nrects = len(left)
+
+        # here comes the tricky part -- we have to set up the vertex and path
+        # codes arrays using moveto, lineto and closepoly
+
+        # for each rect: 1 for the MOVETO, 3 for the LINETO, 1 for the
+        # CLOSEPOLY; the vert for the closepoly is ignored but we still need
+        # it to keep the codes aligned with the vertices
+        nverts = nrects * (1 + 3 + 1)
+        self._verts = np.zeros((nverts, 2))
+        codes = np.ones(nverts, int) * path.Path.LINETO
+        codes[0::5] = path.Path.MOVETO
+        codes[4::5] = path.Path.CLOSEPOLY
+
+        self._verts[0::5, 0] = left
+        self._verts[1::5, 0] = left
+
+        self._verts[2::5, 0] = right
+        self._verts[3::5, 0] = right
+
+        self._verts[1::5, 1] = top
+        self._verts[2::5, 1] = top
+
+        self._verts[0::5, 1] = self._bottom
+        self._verts[3::5, 1] = self._bottom
+
+        barpath = path.Path(self._verts, codes)
+        patch = patches.PathPatch(barpath, facecolor="green", edgecolor="yellow", alpha=0.5)
+        ax.add_patch(patch)
+
+        ax.set_xlim(left[0], right[-1])
+        ax.set_ylim(self._bottom.min(), top.max())
+        self.artists.append(patch)
+
+    def animate(self, frame: int):
+        # simulate new data coming in
+        data = self.data[frame]
+
+        n, bins = np.histogram(data, self.num_bins)
+        top = self._bottom + n
+        self._verts[1::5, 1] = top
+        self._verts[2::5, 1] = top
+
+
+def _make_init_func(plots: List[SwarmPlot], axes: List[plt.Axes]):
+    """create a chained init for each frame"""
+
+    def inner():
+        all_artists = []
+        for p, ax in zip(plots, axes):
+            # print(p.artists)
+            p.init(ax)
+        return all_artists
+
+    return inner
+
+
+def _make_animate_func(plots: List[SwarmPlot]):
+    def inner(frame: int):
+        all_artists = []
+        for p in plots:
+            p.animate(frame)
+            all_artists.extend(p.artists)
+        return all_artists
+
+    return inner
+
+
+def swarm_animate(plots: List[SwarmPlot], destfile: str, num_frames: Optional[int] = None):
+    if len(plots) > 2 or len(plots) < 1:
+        raise ValueError("Must be 1 or 2 plots while under development")
```

Comment:
Ok?  If people abused the flexibility and it broke or presented poorly, I don't think that would be a bad outcome either while under development!

Reply:
Haha, it's more that I literally hardcode the subplot layout. It would explicitly fail right now. I do have a better idea on how to handle it.